src/test/scala/mesosphere/marathon/api/v2/json/V2AppDefinitionTest.scala

```diff
@@ -17,50 +16,58 @@ import play.api.libs.json.{ JsError, Json }
 
 import scala.collection.immutable.Seq
 import scala.concurrent.duration._
+import com.wix.accord._
+import mesosphere.marathon.api.v2.Validation.getAllRuleConstrains
 
 class V2AppDefinitionTest extends MarathonSpec with Matchers {
 
   test("Validation") {
-    def shouldViolate(app: V2AppDefinition, path: String, template: String) = {
-      val violations = ModelValidation.checkAppConstraints(app, PathId.empty)
-      assert(
-        violations.exists { v =>
-          v.getPropertyPath.toString == path && v.getMessageTemplate == template
-        },
-        s"Violations:\n${violations.mkString}"
-      )
+    def shouldViolate(app: V2AppDefinition, path: String, template: String): Unit = {
+      validate(app) match {
+        case Success => fail()
+        case f: Failure =>
+          val violations = getAllRuleConstrains(f)
+          assert(violations.exists { v =>
+            v.property.contains(path) && v.message == template
+          },
+            s"Violations:\n${violations.mkString}"
+          )
+      }
     }
 
-    def shouldNotViolate(app: V2AppDefinition, path: String, template: String) = {
-      val violations = ModelValidation.checkAppConstraints(app, PathId.empty)
-      assert(
-        !violations.exists { v =>
-          v.getPropertyPath.toString == path && v.getMessageTemplate == template
-        },
-        s"Violations:\n${violations.mkString}"
-      )
+    def shouldNotViolate(app: V2AppDefinition, path: String, template: String): Unit = {
+      validate(app) match {
+        case Success =>
+        case f: Failure =>
+          val violations = getAllRuleConstrains(f)
+          assert(!violations.exists { v =>
+            v.property.contains(path) && v.message == template
+          },
+            s"Violations:\n${violations.mkString}"
+          )
+      }
     }
 
     var app = V2AppDefinition(id = "a b".toRootPath)
-    val idError = "path contains invalid characters (allowed: lowercase letters, digits, hyphens, \".\", \"..\")"
+    val idError = "must fully match regular expression '^(([a-z0-9]|[a-z0-9][a-z0-9\\-]*[a-z0-9])\\.)*([a-z0-9]|[a-z0-9][a-z0-9\\-]*[a-z0-9])|(\\.|\\.\\.)$'"
     validateJsonSchema(app, false)
-    shouldViolate(app, "id", idError)
+    shouldViolate(app, "id.path", idError)
```

Comment:
This one was a bit trickier than I thought. There is a `PathId` Validator, which I am calling from `V2AppDefinition` on `id` property. That causes it to be a sub-violation of `V2AppDefinition` validation, which is internally correct. The problem is, that the JSON parser traverses down the violation tree and there is no way to mark `PathId`'s violation descriptions as 'should be ignored'. What I had to do is to extract the validation functionality out of PathId's validator and call it manually on V2AppDefinition's `id` property:
```scala
implicit val appDefinitionValidator = validator[V2AppDefinition] { appDef =>
    appDef.id is validPath
```
`validPath` being a public validation function defined in `PathId`. The nicer former version was:
```scala
implicit val appDefinitionValidator = validator[V2AppDefinition] { appDef =>
    appDef.id is valid
```
which was using the implicit validator for `PathId` classes. Are you fine with that change? You can look it up in `V2AppDefinition` and `PathId` classes of `Extracted PathId validation into a function for external usage. Needed to avoid drilling down the violation chain.` commit.